go/v1beta1/storage/pgsqlstore.go

```diff
@@ -553,27 +553,33 @@ func (pg *PgSQLStore) ListNotes(ctx context.Context, pID, filter, pageToken stri
 		}
 		ns = append(ns, &n)
 	}
-	if count == lastID {
+	if len(ns) == 0 {
+		return ns, "", nil
+	}
+	maxQuery := fmt.Sprintf(notesMaxID, filterQuery)
+	maxID, err := pg.max(ctx, maxQuery, pID)
+	if err != nil {
+		return nil, "", status.Error(codes.Internal, "Failed to query max note id from database")
+	}
+	if lastID >= maxID {
 		return ns, "", nil
 	}
 	encryptedPage, err := encryptInt64(lastID, pg.paginationKey)
 	if err != nil {
-		return nil, "", status.Error(codes.Internal, "Failed to paginate projects")
+		return nil, "", status.Error(codes.Internal, "Failed to paginate notes")
 	}
 	return ns, encryptedPage, nil
 }
 
 // ListNoteOccurrences returns up to pageSize number of occcurrences on the particular note (nID)
 // for this project (pID) projects beginning at pageToken (or from start if pageToken is the empty string).
+// TODO: implement query filter for NoteOccurrences.
```

Comment:
Could you please create an issue and link it here so in the future we can track and identify the improvements we can make more easily?